### DISCUSSION TITLE
projection dimension coordinates to decimal lat/long

### DISCUSSION THREAD

**Discussion by rajkumar8581** [pangeo-data/xESMF#203] [General]
Hello friends, I have an xarray dataset with projection dimension coordinates (meters in x and y directions from the projection origin in the grid mapping space). Could anyone help me with regridding to decimal lat/long.

(y)
float64
2.165e+06 2.195e+06 ... 4.325e+06
units :
m
standard_name :
projection_y_coordinate
axis :
Y
array([2164827.903951, 2194827.903951, 2224827.903951, 2254827.903951,
       2284827.903951, 2314827.903951, 2344827.903951, 2374827.903951,
       2404827.903951, 2434827.903951, 2464827.903951, 2494827.903951,
       2524827.903951, 2554827.903951, 2584827.903951, 2614827.903951,

**Comment by aulemahal** [comment]
Hi @rajkumar8581, 
We would be happy to help, but I think we need more information. 

What's your question exactly? What are you trying to do?

`xESMF` works well for transforming data between different grids, but it is all based on decimal lat-lon information. If your dataset _does not_ have _any_ information about latitude and longitude, then you won't be able to use `xESMF`. You would first need to translate the `x, y` coordinates in `lat(x, y)` and `lon(x,y)` maps. Depending on what information you have, this could be done with `pyproj`.

**Reply by rajkumar8581** [reply]
Thanks for your response @aulemahal, I am basically working with WRF model output files (staggered grids) with coordinates as follows:-
Goal is to simply regrid with decimal lat/long and visualize the variables.

![image](https://user-images.githubusercontent.com/95094560/202015367-efcbe9b5-7d3b-4019-858a-acdb4d7958ba.png)

**Reply by aulemahal** [reply]
I see. So here the issue when using `xESMF` might be that there are multiple lat/lon coordinates, so you must help it find the correct one.

`XLAT`, `XLONG`, `XLAT_U`, `XLONG_U`, `XLAT_V` and `XLONG_V` are all decimal latitude and longitude coordinates. 

Let's say your variable is called 'UVEL` and has dimensions `(y, x_stag)`, this means the correct lat/lon coordinates to associate it to are `XLAT_U`, `XLONG_U`. The simplest way I can see here, without knowing anything about your data, would be to do something like:
```python3
import xesmf as xe

 # This creates a dataset with only 1 data variable and only the coordinates that relate to it
ds_in = ds.UVEL.to_dataset()
# Rename the lat/lon coordinates to lat/lon, so that xESMF knows to use them.
ds_in = ds_in.rename(XLAT_U='lat', XLONG_U='lon')

regridder = xe.Regridder(ds_in, grid_out, method='bilinear')
ds_out = regridder(ds_in)
```
Here, `grid_out` is another dataset with the 1D lat and lon you want.

**Reply by rajkumar8581** [reply]
Hi again Pascal, thanks for the sample code. But I get a NameError for regridder.

regridder = xe.Regridder(ds_in, grid_out, method='bilinear')
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
Cell In [12], line 1
----> 1 regridder = xe.Regridder(ds_in, grid_out, method='bilinear')
      2 ds_out = regridder(ds_in)

NameError: name 'grid_out' is not defined

Secondly, would you be willing to take a look at my data here:-
https://actvet-my.sharepoint.com/:f:/g/personal/rajkumar_sivaprahasam_adpoly_ac_ae/EofPOyhq8UhPlq3VBrvqLJABVR0G8xj_y0gmKRgLW1wagA?e=4hfZuT

**Reply by aulemahal** [reply]
The `NameError: name 'grid_out' is not defined` might be because `grid_out` is not defined.
I think you could read the documentation for `xESMF` to learn how to use it. For example  this page : https://pangeo-xesmf.readthedocs.io/en/latest/notebooks/Curvilinear_grid.html

Good luck!

**Reply by rajkumar8581** [reply]
Got it @aulemahal  Thank you. May I request you to visit xWRF and collaborate if it interests you.

**Comment by raphaeldussin** [comment]
to convert between x,y and lon, lat you can use the pyproj package with the relevant "PROJ" string of your Coordinate Reference System. SInce this is not related to xesmf, can I close the topic?

**Reply by rajkumar8581** [reply]
@raphaeldussin thank you. I was referred to esmf for a possible solution. The dataset has decimal lat long as a staggered grid. I need to cull out 1D lat long from it. @aulemahal
